udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,188 @@
+   * If events do not overlap, no changes will occur.
+   */
+  private Collection<TimeRange> getReplacementForOverlappedTimeRange(
+      TimeRange overlappingTimeRange, TimeRange toReplace) {
+
+    Collection<TimeRange> replacementRanges = new HashSet<>();
+    // Obtain timeRange and overlappingTimeRange start/end data..
+    int toReplaceStartMinute = toReplace.start();
+    int toReplaceEndMinute = toReplace.end();
+
+    int overlappingStartMinute = overlappingTimeRange.start();
+    int overlappingEndMinute = overlappingTimeRange.end();
+
+    int overlapType = classifyOverlap(overlappingTimeRange, toReplace);
+
+    // Case 1&4: Trim front of toReplace.
+    if ((overlapType == OVERLAP_WITH_SAME_START) || (overlapType == OVERLAPS_START)) {
+      TimeRange replaceTime = TimeRange.fromStartDuration(
+          overlappingEndMinute, toReplaceEndMinute - overlappingEndMinute);
+      replacementRanges.add(replaceTime);
+    }
+
+    // Case 2&3: Trim end of toReplace.
+    else if ((overlapType == OVERLAP_WITH_SAME_END) || (overlapType == OVERLAPS_END)) {
+      TimeRange replaceTime = TimeRange.fromStartDuration(
+          toReplaceStartMinute, overlappingStartMinute - toReplaceStartMinute);
+      replacementRanges.add(replaceTime);
+    }
+    
+    // Case 5: Chop overlappingTimeRange out of toReplace.
+    else if (overlapType == OVERLAP_CONTAINS) {
+      TimeRange replaceTimeA = TimeRange.fromStartDuration(
+          toReplaceStartMinute, overlappingStartMinute - toReplaceStartMinute);
+      TimeRange replaceTimeB = TimeRange.fromStartDuration(
+          overlappingEndMinute, toReplaceEndMinute - overlappingEndMinute);
+      replacementRanges.add(replaceTimeA);
+      replacementRanges.add(replaceTimeB);
+    }
+    return replacementRanges;
+  }
+
+  /**
+   * Returns integer representing overlap case type:
+   *  1. OVERLAP_WITH_SAME_START - TimeRanges start at same time and overlappingTimeRange
+   *     ends sooner.
+   *  2. OVERLAP_WITH_SAME_END - overlappingTimeRange starts during timeRange and
+   *     both end together.
+   *  3. OVERLAPS_END - overlappingTimeRange starts during timeRange and ends after.
+   *  4. OVERLAPS_START - overlappingTimeRange starts before timeRange and ends
+   *     during.
+   *  5. OVERLAP_CONTAINS - overlappingTimeRange is completely overlapped by
+   *     timeRange.
+   *  0. Does not overlap
+   */
+  private int classifyOverlap(TimeRange overlappingTimeRange, TimeRange timeRange) {
+    // Obtain timeRange and overlappingTimeRange start/end data..
+    int timeRangeStartMinute = timeRange.start();
+    int timeRangeEndMinute = timeRange.end();
+
+    int overlappingStartMinute = overlappingTimeRange.start();
+    int overlappingEndMinute = overlappingTimeRange.end();
+
+    // Define Overlap Cases
+    boolean overlapWithSameStart = (timeRangeStartMinute == overlappingStartMinute)
+        && (timeRangeEndMinute > overlappingEndMinute);
```

Comment:
This would be clearer:

```
if ((timeRangeStartMinute == overlappingStartMinute)
        && (timeRangeEndMinute > overlappingEndMinute)) {
  return OVERLAP_WITH_SAME_START;
}

if (...) {
  return OVERLAP_WITH_SAME_END;
}

...
```

Reply:
Done